### DISCUSSION TITLE
How to setup Nuxt3 into Codespaces with HMR?

### DISCUSSION THREAD

**Discussion by NBaron** [nuxt/nuxt#16199] [Questions]
Hello all,
Nuxt 3.0.0 is almost working in GitHub Codespaces.
**Only remaining issue is the Vite HMR setup not able to connect via WebSocket.**

.
### _2) How to setup Vite port in Nuxt3 (not Nuxt port)?_
.

### Repro
With a basic nuxt app HMR is not working out of the box.

Due to port forwarding of Codespaces and HTTPS, these errors pop up:
`client:243 Mixed Content: The page at 'https://foo-bar-train-baz6x47w5vfj695-3000.preview.app.github.dev/' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://foo-bar-train-baz6x47w5vfj695-3000.preview.app.github.dev:24678/_nuxt/'. This request has been blocked; this endpoint must be available over WSS.`

and

`[vite] failed to connect to websocket (SecurityError: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.). `

So, HMR requires a specific setup for Codespaces. After a few (long hours) of search, here is the nuxt.config.ts:
```
const hmr = process.env.CODESPACES
? {
    host: process.env['CODESPACE_NAME'] + '-24678.' + process.env['GITHUB_CODESPACES_PORT_FORWARDING_DOMAIN'],
    port: 24678,
    clientPort: 443,
    protocol: "wss",
}
: {
    host: 'localhost',
    port: 24678,
    clientPort: 24678,
};

export default defineNuxtConfig({
    vite: {
        server: {
            hmr,
        }
    },
    // ...truncated...
})
```

This is an improved code based on https://github.com/vitejs/vite/pull/677#issuecomment-1169387835

The remaining issue is that `yarn dev` now produces an error related to WebSocket:
````
 ERROR  WebSocket server error:                                                                                                                                                                                 23:31:53
Error: listen EADDRNOTAVAIL: address not available 20.71.22.126:24678
    at Server.setupListenHandle [as _listen2] (node:net:1446:21)
    at listenInCluster (node:net:1511:12)
    at GetAddrInfoReqWrap.doListen [as callback] (node:net:1660:7)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:111:8)
````

To solve this issue, here https://github.com/vitejs/vite/pull/677#issuecomment-1169387835, it is suggested to force vite on port 24678 with:
`vite --port 24678`

I tried `yarn dev --port 24678` but this set the port for nuxt and not for vite.
I don't know how to achieve this with Nuxt3. I need your help.

Thank you in advance.

Potentially related to:
https://github.com/nuxt/framework/discussions/8178
https://github.com/nuxt/framework/discussions/9221

**Comment by fuzzzerd** [comment]
Seeing this is an older post, wondering if anyone has found a way to get this working?

**Reply by maou-shonen** [reply]
Have you tried the port 24678 mentioned above?
```ts
export default defineNuxtConfig({
  vite: {
    server: {
      hmr: {
        port: 24678 
      }
    }
  }
}
```

**Reply by NBaron** [reply]
I have to admit that I abandoned the idea (for the moment).
I tried different things as you can see, but I did not had a look since the latest updates after 3.0.0.

Perhaps there is a new config option to achieve this.

I am still looking forward at a solution for this.

**Comment by enricopolanski** [comment]
Any update? This is kinda blocking for our team looking to get onto devcontainers as we have 6 different nuxt projects.

**Comment by jacobkuzmits** [comment]
I have a seemingly working but hacky solution, although I'm not sure I understand -why- it works. After reading through many discussions on the issue and trying a lot of things, I stumbled on this.


```
nuxt.config.ts


export default defineNuxtConfig({
    vite: {
        plugins: [
            {
                name: "codespaces-hmr-fix",
                transform(code, id) {
                    if (id.endsWith("dist/client/client.mjs") || id.endsWith("dist/client/env.mjs")) {
                        return code.replace(
                            "`${__HMR_HOSTNAME__ || importMetaUrl.hostname}:${hmrPort || importMetaUrl.port}${__HMR_BASE__}`",
                            JSON.stringify("silver-meme-vxw967xxj93xq5v-5175.app.github.dev:443/_nuxt/")
                        );
                    }
                    return code;
                },
            },
        ],
        server: {
            hmr: {
                clientPort: 443,
                port: 5175,
                protocol: "wss",
            },
        },
    },
});
```

Obviously the "silver-meme-vxw967xxj93xq5v-" portion of the URL is unique to each codespace and the port being 5175 is completely arbitrary. After setting this config and running `npm run dev`, Codespaces prompts me to open both port 3000 and 5175 in the browser, and in the ports tab they are both forwarded. If you open port 3000, the websocket will still refuse to connect and constantly reload. But once you open port 5175 in the browser, which simply shows "Upgrade required", it stops reloading and HMR starts working. After restarting the codespace, it does seem that you have to repeat opening the Vite port in the browser, which is less than ideal. 

I know this can be improved with environment variables and the such, but I just wanted to get this out there. Maybe someone else can turn this into a less hacky fix that doesn't require opening a 2nd tab each time the codespace starts.

**Reply by trylovetom** [reply]
great, it's work on my proxy server

**Comment by coderhammer** [comment]
I had a lot of problems but I tried something randomly and it looks like it works pretty great. Here is my configuration:

```ts
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  devtools: { enabled: true },
  ssr: false,
  vite: {
    plugins: [],
    server: {
      middlewareMode: true,
    },
  },
});
```

**Reply by NBaron** [reply]
> I had a lot of problems but I tried something randomly and it looks like it works pretty great. Here is my configuration:
> 
> ```ts
> // https://nuxt.com/docs/api/configuration/nuxt-config
> export default defineNuxtConfig({
>   devtools: { enabled: true },
>   ssr: false,
>   vite: {
>     plugins: [],
>     server: {
>       middlewareMode: true,
>     },
>   },
> });
> ```

Perhaps the `vite.server.middlewareMode` is doing the magic.
I'll look into this.

**Comment by vonec** [comment]
```
export default defineNuxtConfig({

  vite: {
    server: {
      hmr: {
        protocol: process.env.CODESPACES ? 'wss' : 'ws',
        host: process.env.CODESPACE_NAME ? `${process.env.CODESPACE_NAME}-3000.githubpreview.dev` : 'localhost',
        port: 3000
      }
    } 
  },
});
```

This worked for me

**Reply by sugoidesune** [reply]
I tried on a clean nuxt 3.13 install and this made it work!
No other changes needed.

**Reply by peterhirn** [reply]
Unlikely that this works as expected:

- githubpreview.dev is obsolete (see https://github.blog/changelog/2022-10-20-introducing-the-codespaces-simple-browser/)
- vite can't bind the websocket to this host
- websocket can't be on the same port as the dev server

 Your page is probably crashing and reloading continuously, which looks like HMR? 😅

**Reply by sugoidesune** [reply]
Task failed successfully haha. But I dont see the page constantly reloading when there are no changes.
and when i make a change i see this in the console:
`[vite] hot updated: /app.vue`

**Reply by peterhirn** [reply]
Maybe I'm completely wrong 😁 If it works for you correctly that's great.
